File: 2_term/2_C#/2-3/2-3/Properties/List.cs
using System;

namespace Application
{
    public class List
    {
        public void Add(int value)
        {
            head = new ListElement(value, head);
            length++;
        }

        public void Add(int value, int index)
        {
            if (index != 0)
            {
                ListElement current = head;
                for (int i = 0; i < index - 1; i++)
                {
                    current = current.LNext;
                }
                current.LNext = new ListElement(value, current.LNext);
                length++;
            }
            else
            {
                Add(value);
            }
        }

        public int Index(int index)
        {
            ListElement current = head;
            for (int i = 0; i < index; i++)
            {
                current = current.LNext;
            }
            return current.Value;
        }

        public bool RemoveValue(int value)
        {
            ListElement current = head;
            ListElement previous = head;
            while (current != null && current.Value != value)
            {
                previous = current;
                current = current.LNext;
            }
            if (current != null)
            {
                if (current != head)
                    previous.LNext = current.LNext;
                else
                    head = current.LNext;
                length--;
                return true;
            }
            return false;
        }

        public bool RemoveIndex(int index)
        {
            if (!IsExistsIndex(index))
            {
                return false;
            }
            if (index != 0)
            {
                ListElement current = head;
                for (int i = 0; i < index - 1; i++)
                {
                    current = current.LNext;
                }
                current.LNext = current.LNext.LNext;
            }
            else
            {
                head = head.LNext;
            }
            length--;
            return true;
        }

        private bool IsExistsIndex(int index)
        {
            if (head == null)
                return false;
            ListElement current = head;
            for (int i = 0; i < index; i++)
            {
                if (current.LNext == null)
                    return false;
                current = current.LNext;
            }
            return true;
        }

        public bool IsInList(int value)
        {
            ListElement current = head;
            while (current != null && current.Value != value)
            {
                current = current.LNext;
            }
            if (current != null)
            {
                return true;
            }
            return false;
        }

        public void Print()
        {
            ListElement current = head;
            while (current != null)
            {
                Console.Write("{0} ", current.Value);
                current = current.LNext;
            }
        }

        public int Length()
        {
            return length;
        }

        private class ListElement
        {
            public int Value { set; get; }

            public ListElement(int value, ListElement lNext)
            {
                this.LNext = lNext;
                this.Value = value;
            }

            public ListElement LNext { set; get; }
        };

        private ListElement head = null;
        private int length = 0;
    };
}

File: 2_term/7/7-2/7-2/list.cpp
#include "list.h"


List::List() : lengthField(0), head(nullptr)
{
}

List::~List()
{
    ListElement* current = head;
    while (current != nullptr)
    {
        ListElement* toDelete = current;
        current = current->lNext;
        delete toDelete;
    }
    lengthField = 0;
}

void List::add(int value)
{
    head = new ListElement(value, head);
    lengthField++;
}

void List::add(int value, int index)
{
    if (index != 0)
    {
        ListElement* current = head;
        for (int i = 0; i < index - 1; i++)
        {
            current = current->lNext;
        }
        current->lNext = new ListElement(value, current->lNext);
        lengthField++;
    }
    else
    {
        add(value);
    }
}

int List::index(int index)
{
    ListElement* current = head;
    for (int i = 0; i < index; i++)
    {
        current = current->lNext;
    }
    return current->value;
}

void List::removeValue(int value)
{
    ListElement* current = head;
    ListElement* previous = head;
    while (current != nullptr && current->value != value)
    {
        previous = current;
        current = current->lNext;
    }
    if (current != nullptr)
    {
        if (current != head)
            previous->lNext = current->lNext;
        else
            head = current->lNext;
        lengthField--;
    }
}

void List::removeIndex(int index)
{
    if (index != 0)
    {
        ListElement* current = head;
        for (int i = 0; i < index - 1; i++)
        {
            current = current->lNext;
        }
        current->lNext = current->lNext->lNext;
    }
    else
    {
        head = head->lNext;
    }
    lengthField--;
}

bool List::isExistsIndex(int index)
{
    return (lengthField >= index);
}

bool List::isInList(int value)
{
    ListElement* current = head;
    while (current != nullptr && current->value != value)
    {
        current = current->lNext;
    }
    return current != nullptr;
}

void List::print()
{
    ListElement* current = head;
    while (current != nullptr)
    {
        std::cout << current->value << " ";
        current = current->lNext;
    }
}

int List::length()
{
    return lengthField;
}
